File: solutions/java/relative-distance/1/src/main/java/RelativeDistance.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class RelativeDistance {

    private final Map<String, List<String>> familyTree;
    private Map<String, Map<String, Integer>> distanceCache;

    RelativeDistance(Map<String, List<String>>  familyTree) {
        this.familyTree = generateFullFamilyTree(familyTree);
        this.distanceCache = new java.util.HashMap<String, Map<String, Integer>>();
    }

    private Map<String, List<String>> generateFullFamilyTree(Map<String, List<String>> familyTree) {
        Map<String, List<String>> fullFamilyTree = new HashMap<String, List<String>>();
        // Add all relationships from the original family tree
        for (String person : familyTree.keySet()) {
            fullFamilyTree.putIfAbsent(person, new ArrayList<>());
            fullFamilyTree.get(person).addAll(familyTree.get(person));
        }
        // Make relationships bidirectional chid <-> parent
        for (String person : familyTree.keySet()) {
            List<String> relatives = familyTree.get(person);
            for (String relative : relatives) {
                fullFamilyTree.putIfAbsent(relative, new ArrayList<>());
                if (!fullFamilyTree.get(relative).contains(person)) {
                    fullFamilyTree.get(relative).add(person);
                }
            }
        }
        //  make relationships bidirectional sibling <-> sibling
        for (String person : familyTree.keySet()) {
            List<String> relatives = familyTree.get(person);
            for (String relative : relatives) {
                for (String sibling : relatives) {
                    if (sibling.equals(relative)) { continue; }
                    fullFamilyTree.putIfAbsent(sibling, new ArrayList<>());
                    // Avoid adding a sibling to their own list or to a 
                    // relative's list if already present
                    if (!fullFamilyTree.get(relative).contains(sibling)) {
                        fullFamilyTree.get(relative).add(sibling);
                    }
                }
            } 
        }
        return fullFamilyTree;
    }

    int degreeOfSeparation(String personA, String personB) {
        return degreeOfSeparation(personA, personB, new ArrayList<String>());
    }

    private int degreeOfSeparation(String personA, String personB, List<String> visited) {
        if (personA.equals(personB)) { return 0; }

        // cache lookup
        if (distanceCache.containsKey(personA)) {
            Map<String, Integer> cachedDistances = distanceCache.get(personA);
            if (cachedDistances.containsKey(personB)) {
                return cachedDistances.get(personB);
            }
        }
        else {
            distanceCache.putIfAbsent(personA, new HashMap<>()); 
        }
        if (distanceCache.containsKey(personB)) {
            Map<String, Integer> cachedDistances = distanceCache.get(personB);
            if (cachedDistances.containsKey(personA)) {
                return cachedDistances.get(personA);
            }
        }
        else {
            distanceCache.putIfAbsent(personB, new HashMap<>());
        }

        List<String> relatives = familyTree.get(personA);
        
        if (relatives == null || relatives.isEmpty()) {
            distanceCache.get(personA).put(personB, -1);
            return -1;
        }

        int minDistance = -1; // family.size() + 1;
        boolean found = false;
        for (String relative : relatives) {
            if (visited.contains(relative)) { continue; }
                        
            List<String> newVisited = new ArrayList<>(visited);
            newVisited.add(relative);
            
            int distance = degreeOfSeparation(relative, personB, newVisited);
            if (distance == -1) { continue; }
            distance += 1; // account for the step to the relative
            if (!found || distance < minDistance) {
                minDistance = distance;
                found = true;
            }
        }
        
        // Store in cache (both directions)
        distanceCache.get(personA).put(personB, minDistance);
        distanceCache.get(personB).put(personA, minDistance);
        // If no relatives found, return -1
        return minDistance;
    }
}

File: solutions/python/relative-distance/1/relative_distance.py
"""
# Relative Distance
determine the degree of separation between two individuals in a family tree. 
"""

from collections import defaultdict, deque


class RelativeDistance:
    """
    Compute the degree of separation between individuals in a family tree.

    The family tree is represented as an undirected graph where:
    - Parent-child relationships have a distance of 1
    - Sibling relationships have a distance of 1

    The class builds this graph and allows querying the shortest distance
    (degree of separation) between any two individuals using BFS.
    """

    _family_graph = defaultdict(set)

    def __init__(self, family_tree):
        """
        Initialize the family graph from a given family tree structure.

        Args:
            family_tree (dict):
                A dictionary mapping each parent (str) to a list of their children (list[str]).

        Behavior:
            - Adds bidirectional edges between parents and children.
            - Adds bidirectional edges between siblings.
        """
        for parent, children in family_tree.items():
            for i, child in enumerate(children):
                self._family_graph[parent].add(child)
                self._family_graph[child].add(parent)

                # Connect siblings
                for child_j in children[i + 1:]:
                    self._family_graph[child].add(child_j)
                    self._family_graph[child_j].add(child)

    def degree_of_separation(self, person_a, person_b):
        """
        Calculate the minimum degree of separation between two individuals.

        This uses Breadth-First Search (BFS) to find the shortest path
        between the two nodes in the graph.

        Args:
            person_a (str): The starting individual.
            person_b (str): The target individual.

        Returns:
            int: The number of edges in the shortest path between person_a and person_b.

        Raises:
            ValueError:
                - If person_a is not in the family tree.
                - If person_b is not in the family tree.
                - If there is no connection between the two individuals.
        """
        if person_a not in self._family_graph:
            raise ValueError("Person A not in family tree.")
        if person_b not in self._family_graph:
            raise ValueError("Person B not in family tree.")

        visited = set()
        queue = deque([(person_a, 0)])

        while queue:
            current, dist = queue.popleft()

            if current == person_b:
                return dist

            if current in visited:
                continue

            visited.add(current)

            for neighbor in self._family_graph[current]:
                if neighbor not in visited:
                    queue.append((neighbor, dist + 1))

        raise ValueError("No connection between person A and person B.")
